3Dmol/glmodel.js

```diff
@@ -1811,36 +1811,59 @@ $3Dmol.GLModel = (function() {
                 return true; // undef gets all
             var invert = !!sel.invert;
             var ret = true;
-            for ( var key in sel) {
-                if(key == "and" || key == "or" || key == "not"){//boolean fields
-                    if(key == "not"){
-                        if(this.atomIsSelected(atom,sel[key])){
-                            ret=false;
+            for (var key in sel) {
+                if (key == "and" || key == "or" || key == "not") {  //boolean operators
+                    if (key == "not") {
+                        if (this.atomIsSelected(atom, sel[key])) {
+                            ret = false;
                             break;
                         }
-                    }else{//"or" and "and"
-                        if(key == "and"){
-                            var and=sel[key];//array format
-                            for(var i=0;i<and.length;i++){
-                                if(!this.atomIsSelected(atom,and[i])){
-                                    ret=false;
-                                    break;
+                    } else { //"or" and "and"
+                        // cache the evaluation of the selection for better performance
+                        if (sel[key].__cached_results === undefined) {
```

Comment:
> It would also be a good idea to memoize the result of the selectedAtoms call for even better efficiency.

This is how I've implemented it, caching the full evaluation of the and/or part of the selection. My only concern here is that the selection could change without the code realizing with something like

```js
const selection = {and: [{serial: 4}, {resn: "ALA"}]};
model.selectedAtoms(selection);

// the selection changes, but we don't know it!
selection.and[0].serial = 67;
model.selectedAtoms(selection);
```

I can see two ways to solve this: add the hash of the selection with the cached result (something like `md5(JSON.stringify(sel))`, and check that this did not change since last evaluation; or clean/remove `__cached_results` when starting a new evaluation of the selection.

The second solution require that there is a single entry point for selection evaluation, which could be `GLModel.selectedAtoms`. Currently, it looks like users could also use `atomIsSelected` as an entry point, so I could make this a private function.

As a side note, it looks like `atomIsSelected` do not accept the same selection elements as `selectedAtoms` (stuff like within/byres is not supported). Should this function be made private anyway?

Reply:
Rather than hashing, my preference would be to copy the initial selection object.  At little more efficient computationally, a little less efficient memory-wise.  We would need to write a helper function that recursively traverses the selection object copying fields as needed (no need to copy if a cache already exists).  It could also precompute the cached selections.  Then instead of atomIsSelected computing the selections, it will check for the existence of a cache, if it exists it uses it, otherwise it does what it does now.  This keeps backwards compatibility but makes the common path more efficient.

What do you think?

Reply:
Where would the copy of the selection object take place? In `selectedAtoms` or in `atomIsSelected`? I guess `selectedAtoms` would make the most sense here.